Modules/pyexpat.c

```diff
@@ -1632,147 +1632,188 @@ static int init_handler_descrs(void)
+    if (mod == NULL) {
         return NULL;
+    }
 
-    if (PyType_Ready(&Xmlparsetype) < 0 || init_handler_descrs() < 0)
-        return NULL;
+    PyObject *codes_dict = NULL, *rev_codes_dict = NULL;
 
-    /* Create the module and add the functions */
-    m = PyModule_Create(&pyexpatmodule);
-    if (m == NULL)
-        return NULL;
+    if (PyType_Ready(&Xmlparsetype) < 0) {
+        goto error;
+    }
+
+    if (init_handler_descrs() < 0) {
+        goto error;
+    }
 
     /* Add some symbolic constants to the module */
     if (ErrorObject == NULL) {
         ErrorObject = PyErr_NewException("xml.parsers.expat.ExpatError",
-                                         NULL, NULL);
-        if (ErrorObject == NULL)
-            return NULL;
+                                        NULL, NULL);
+    }
+    if (ErrorObject == NULL) {
+        goto error;
     }
+
     Py_INCREF(ErrorObject);
-    PyModule_AddObject(m, "error", ErrorObject);
+    if (PyModule_AddObject(mod, "error", ErrorObject) < 0) {
+        Py_DECREF(ErrorObject);
+        goto error;
+    }
     Py_INCREF(ErrorObject);
-    PyModule_AddObject(m, "ExpatError", ErrorObject);
+    if (PyModule_AddObject(mod, "ExpatError", ErrorObject) < 0) {
+        Py_DECREF(ErrorObject);
+        goto error;
+    }
     Py_INCREF(&Xmlparsetype);
-    PyModule_AddObject(m, "XMLParserType", (PyObject *) &Xmlparsetype);
+    if (PyModule_AddObject(mod, "XMLParserType",
+                           (PyObject *) &Xmlparsetype) < 0) {
+        Py_DECREF(&Xmlparsetype);
+        goto error;
+    }
 
-    PyModule_AddStringConstant(m, "EXPAT_VERSION",
-                               XML_ExpatVersion());
+    if (PyModule_AddStringConstant(mod, "EXPAT_VERSION",
+                                   XML_ExpatVersion()) < 0) {
+        goto error;
+    }
     {
         XML_Expat_Version info = XML_ExpatVersionInfo();
-        PyModule_AddObject(m, "version_info",
-                           Py_BuildValue("(iii)", info.major,
-                                         info.minor, info.micro));
+        PyObject *versionInfo = Py_BuildValue("(iii)",
+                                              info.major,
+                                              info.minor,
+                                              info.micro);
+        if (PyModule_AddObject(mod, "version_info", versionInfo) < 0) {
+            Py_DECREF(versionInfo);
+            goto error;
+        }
     }
     /* XXX When Expat supports some way of figuring out how it was
        compiled, this should check and set native_encoding
        appropriately.
     */
-    PyModule_AddStringConstant(m, "native_encoding", "UTF-8");
+    if (PyModule_AddStringConstant(mod, "native_encoding", "UTF-8") < 0) {
+        goto error;
+    }
 
-    d = PyModule_GetDict(m);
-    if (d == NULL) {
-        Py_DECREF(m);
-        return NULL;
+    PyObject *errors_module = PyModule_New(MODULE_NAME ".errors");
```

Comment:
Can you please create a subfunction which creates the .errors submodule?

Reply:
@vstinner sure - same thing for the models submodule too?

Reply:
Yes.